File: Ocis.Server/Ocis.Protocol/ocis-protocol-dart/Library.dart
// ignore_for_file: camel_case_types, constant_identifier_names, non_constant_identifier_names, unnecessary_this
import '../ProtocolSpec.dart' as protocol_spec;
import './fable_modules/fable_library/Array.dart' as array;
import './fable_modules/fable_library/BitConverter.dart' as bit_converter;
import './fable_modules/fable_library/Encoding.dart' as encoding;
import './fable_modules/fable_library/String.dart' as string;
import './fable_modules/fable_library/Types.dart' as types;
import './fable_modules/fable_library/Util.dart' as util;

protocol_spec.RequestPacket CreateRequest(int commandType, List<int> key, types.Some<List<int>>? value) {
    final valueLen = (value == null) ? 0 : types.value(value).length;
    return protocol_spec.RequestPacket(1397310287, 1, commandType, (18 + key.length) + valueLen, key.length, valueLen, key, value);
}

List<int> SerializeRequest(protocol_spec.RequestPacket packet) {
    final parts = <List<int>>[];
    parts.add(bit_converter.getBytesUInt32(packet.MagicNumber));
    parts.add([packet.Version]);
    parts.add([packet.CommandType]);
    parts.add(bit_converter.getBytesInt32(packet.TotalPacketLength));
    parts.add(bit_converter.getBytesInt32(packet.KeyLength));
    parts.add(bit_converter.getBytesInt32(packet.ValueLength));
    parts.add(packet.Key);
    final types.Some<List<int>>? matchValue = packet.Value;
    if (matchValue == null) {
    } else {
        parts.add(types.value(matchValue));
    }
    return array.concat<int>(parts.sublist(0));
}

class ParseResult$1<$T> implements types.Union, Comparable<ParseResult$1<$T>> {
    final int tag;
    const ParseResult$1(this.tag);
    @override
    bool operator ==(Object other) => (other is ParseResult$1<$T>) && (other.tag == tag);
    @override
    int get hashCode => tag.hashCode;
    @override
    int compareTo(ParseResult$1<$T> other) => tag.compareTo(other.tag);
}

class ParseResult$1_ParseSuccess<$T> extends ParseResult$1<$T> {
    final $T Item;
    const ParseResult$1_ParseSuccess(this.Item): super(0);
    @override
    bool operator ==(Object other) => (other is ParseResult$1_ParseSuccess<$T>) && util.equalsDynamic(other.Item, Item);
    @override
    int get hashCode => util.combineHashCodes([tag.hashCode, Item.hashCode]);
    @override
    int compareTo(ParseResult$1<$T> other) {
        if (other is ParseResult$1_ParseSuccess<$T>) {
            return util.compareDynamic(Item, other.Item);
        } else {
            return tag.compareTo(other.tag);
        }
    }
}

class ParseResult$1_ParseError<$T> extends ParseResult$1<$T> {
    final String Item;
    const ParseResult$1_ParseError(this.Item): super(1);
    @override
    bool operator ==(Object other) => (other is ParseResult$1_ParseError<$T>) && (other.Item == Item);
    @override
    int get hashCode => util.combineHashCodes([tag.hashCode, Item.hashCode]);
    @override
    int compareTo(ParseResult$1<$T> other) {
        if (other is ParseResult$1_ParseError<$T>) {
            return Item.compareTo(other.Item);
        } else {
            return tag.compareTo(other.tag);
        }
    }
}

ParseResult$1<protocol_spec.ResponsePacket> DeserializeResponse(List<int> buffer) {
    try {
        if (buffer.length < 18) {
            return const ParseResult$1<protocol_spec.ResponsePacket>(/* InsufficientData */ 2);
        } else {
            var offset = 0;
            final magicNumber = bit_converter.toUInt32(buffer, offset);
            offset = offset + 4;
            final version = buffer[offset];
            offset = offset + 1;
            final statusCode = buffer[offset];
            offset = offset + 1;
            final totalPacketLength = bit_converter.toInt32(buffer, offset);
            offset = offset + 4;
            final valueLength = bit_converter.toInt32(buffer, offset);
            offset = offset + 4;
            final errorMessageLength = bit_converter.toInt32(buffer, offset);
            offset = offset + 4;
            if (buffer.length < totalPacketLength) {
                return const ParseResult$1<protocol_spec.ResponsePacket>(/* InsufficientData */ 2);
            } else if (!((magicNumber == 1397310287) && (version == 1))) {
                return const ParseResult$1_ParseError<protocol_spec.ResponsePacket>('invalid header');
            } else if ((valueLength < 0) || (errorMessageLength < 0)) {
                return const ParseResult$1_ParseError<protocol_spec.ResponsePacket>('invalid length field');
            } else if (totalPacketLength != ((18 + valueLength) + errorMessageLength)) {
                return const ParseResult$1_ParseError<protocol_spec.ResponsePacket>('packet length mismatch');
            } else {
                final types.Some<List<int>>? value = (valueLength > 0) ? types.Some(array.getSubArray<int>(buffer, offset, valueLength)) : null;
                offset = offset + valueLength;
                final types.Some<String>? tmp_capture_2;
                if (errorMessageLength > 0) {
                    final errorBytes = array.getSubArray<int>(buffer, offset, errorMessageLength);
                    tmp_capture_2 = types.Some(encoding.get_UTF8().getString(errorBytes));
                } else {
                    tmp_capture_2 = null;
                }
                return ParseResult$1_ParseSuccess<protocol_spec.ResponsePacket>(protocol_spec.ResponsePacket(magicNumber, version, statusCode, totalPacketLength, valueLength, errorMessageLength, value, tmp_capture_2));
            }
        }
    } catch (ex) {
        final arg = ex.toString();
        return ParseResult$1_ParseError<protocol_spec.ResponsePacket>((string.toText(string.printf('error parsing response: %s')))(arg));
    }
}

List<int> ProtocolHelper_stringToBytes(String s) => encoding.get_UTF8().getBytes(s);

String ProtocolHelper_bytesToString(List<int> bytes) => encoding.get_UTF8().getString(bytes);

protocol_spec.RequestPacket ProtocolHelper_createSetRequest(String key, String value) => CreateRequest(1, ProtocolHelper_stringToBytes(key), types.Some(ProtocolHelper_stringToBytes(value)));

protocol_spec.RequestPacket ProtocolHelper_createGetRequest(String key) => CreateRequest(2, ProtocolHelper_stringToBytes(key), null);

protocol_spec.RequestPacket ProtocolHelper_createDeleteRequest(String key) => CreateRequest(3, ProtocolHelper_stringToBytes(key), null);

File: Ocis.Server/Ocis.Protocol/ocis-protocol-ts/Library.ts
import { uint32, uint8, int32 } from "./fable_modules/fable-library-ts.4.25.0/Int32.js";
import { Option, value as value_1 } from "./fable_modules/fable-library-ts.4.25.0/Option.js";
import { ResponsePacket, RequestPacket } from "../ProtocolSpec.js";
import { toInt32, toUInt32, getBytesInt32, getBytesUInt32 } from "./fable_modules/fable-library-ts.4.25.0/BitConverter.js";
import { getSubArray, item, concat } from "./fable_modules/fable-library-ts.4.25.0/Array.js";
import { Union } from "./fable_modules/fable-library-ts.4.25.0/Types.js";
import { union_type, string_type, TypeInfo } from "./fable_modules/fable-library-ts.4.25.0/Reflection.js";
import { printf, toText } from "./fable_modules/fable-library-ts.4.25.0/String.js";
import { get_UTF8 } from "./fable_modules/fable-library-ts.4.25.0/Encoding.js";

/**
 * create request packet
 */
export function CreateRequest(commandType: int32, key: uint8[], value: Option<uint8[]>): RequestPacket {
    let valueLen: int32;
    if (value == null) {
        valueLen = 0;
    }
    else {
        const v: uint8[] = value_1(value);
        valueLen = v.length;
    }
    return new RequestPacket(1397310287, 1, commandType, (18 + key.length) + valueLen, key.length, valueLen, key, value);
}

/**
 * serialize request packet to bytes
 */
export function SerializeRequest(packet: RequestPacket): uint8[] {
    const parts: uint8[][] = [];
    void (parts.push(Array.from(getBytesUInt32(packet.MagicNumber))));
    void (parts.push([packet.Version]));
    void (parts.push([packet.CommandType & 0xFF]));
    void (parts.push(Array.from(getBytesInt32(packet.TotalPacketLength))));
    void (parts.push(Array.from(getBytesInt32(packet.KeyLength))));
    void (parts.push(Array.from(getBytesInt32(packet.ValueLength))));
    void (parts.push(packet.Key));
    const matchValue: Option<uint8[]> = packet.Value;
    if (matchValue == null) {
    }
    else {
        const value: uint8[] = value_1(matchValue);
        void (parts.push(value));
    }
    return concat<uint8>(parts.slice());
}

export type ParseResult$1_$union<T> = 
    | ParseResult$1<T, 0>
    | ParseResult$1<T, 1>
    | ParseResult$1<T, 2>

export type ParseResult$1_$cases<T> = {
    0: ["ParseSuccess", [T]],
    1: ["ParseError", [string]],
    2: ["InsufficientData", []]
}

export function ParseResult$1_ParseSuccess<T>(Item: T) {
    return new ParseResult$1<T, 0>(0, [Item]);
}

export function ParseResult$1_ParseError<T>(Item: string) {
    return new ParseResult$1<T, 1>(1, [Item]);
}

export function ParseResult$1_InsufficientData<T>() {
    return new ParseResult$1<T, 2>(2, []);
}

export class ParseResult$1<T, Tag extends keyof ParseResult$1_$cases<T>> extends Union<Tag, ParseResult$1_$cases<T>[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: ParseResult$1_$cases<T>[Tag][1]) {
        super();
    }
    cases() {
        return ["ParseSuccess", "ParseError", "InsufficientData"];
    }
}

export function ParseResult$1_$reflection(gen0: TypeInfo): TypeInfo {
    return union_type("Ocis.Protocol.ParseResult`1", [gen0], ParseResult$1, () => [[["Item", gen0]], [["Item", string_type]], []]);
}

/**
 * deserialize response from bytes
 */
export function DeserializeResponse(buffer: uint8[]): ParseResult$1_$union<ResponsePacket> {
    let arg: string, errorBytes: uint8[];
    try {
        if (buffer.length < 18) {
            return ParseResult$1_InsufficientData<ResponsePacket>();
        }
        else {
            let offset = 0;
            const magicNumber: uint32 = toUInt32(buffer, offset);
            offset = ((offset + 4) | 0);
            const version: uint8 = item(offset, buffer);
            offset = ((offset + 1) | 0);
            const statusCode: uint8 = item(offset, buffer);
            offset = ((offset + 1) | 0);
            const totalPacketLength: int32 = toInt32(buffer, offset);
            offset = ((offset + 4) | 0);
            const valueLength: int32 = toInt32(buffer, offset);
            offset = ((offset + 4) | 0);
            const errorMessageLength: int32 = toInt32(buffer, offset);
            offset = ((offset + 4) | 0);
            if (buffer.length < totalPacketLength) {
                return ParseResult$1_InsufficientData<ResponsePacket>();
            }
            else if (!((magicNumber === 1397310287) && (version === 1))) {
                return ParseResult$1_ParseError<ResponsePacket>("invalid header");
            }
            else if ((valueLength < 0) ? true : (errorMessageLength < 0)) {
                return ParseResult$1_ParseError<ResponsePacket>("invalid length field");
            }
            else if (totalPacketLength !== ((18 + valueLength) + errorMessageLength)) {
                return ParseResult$1_ParseError<ResponsePacket>("packet length mismatch");
            }
            else {
                const value: Option<uint8[]> = (valueLength > 0) ? getSubArray<uint8>(buffer, offset, valueLength) : undefined;
                offset = ((offset + valueLength) | 0);
                return ParseResult$1_ParseSuccess<ResponsePacket>(new ResponsePacket(magicNumber, version, statusCode, totalPacketLength, valueLength, errorMessageLength, value, (errorMessageLength > 0) ? ((errorBytes = getSubArray<uint8>(buffer, offset, errorMessageLength), get_UTF8().getString(errorBytes))) : undefined));
            }
        }
    }
    catch (ex: any) {
        return ParseResult$1_ParseError<ResponsePacket>((arg = ex.message, toText(printf("error parsing response: %s"))(arg)));
    }
}

/**
 * convert string to bytes
 */
export function ProtocolHelper_stringToBytes(s: string): uint8[] {
    return Array.from(get_UTF8().getBytes(s));
}

/**
 * convert bytes to string
 */
export function ProtocolHelper_bytesToString(bytes: uint8[]): string {
    return get_UTF8().getString(bytes);
}

/**
 * create SET request
 */
export function ProtocolHelper_createSetRequest(key: string, value: string): RequestPacket {
    return CreateRequest(1, ProtocolHelper_stringToBytes(key), ProtocolHelper_stringToBytes(value));
}

/**
 * create GET request
 */
export function ProtocolHelper_createGetRequest(key: string): RequestPacket {
    return CreateRequest(2, ProtocolHelper_stringToBytes(key), undefined);
}

/**
 * create DELETE request
 */
export function ProtocolHelper_createDeleteRequest(key: string): RequestPacket {
    return CreateRequest(3, ProtocolHelper_stringToBytes(key), undefined);
}

